torch/csrc/distributed/autograd/context/dist_autograd_container.cpp

```diff
@@ -33,35 +41,69 @@ DistAutogradContainer& DistAutogradContainer::getInstance() {
   return container;
 }
 
-const DistAutogradContext& DistAutogradContainer::newContext() {
-  if (!initialized_) {
-    throw std::runtime_error(
-        "Need to initialize distributed autograd using "
-        "torch.distributed.autograd.init()");
+int64_t DistAutogradContainer::newAutogradMessageId() {
+  checkInitialized();
+
+  // Check for int64_t max overflow.
+  TORCH_INTERNAL_ASSERT(
+      next_autograd_message_id_ < std::numeric_limits<int64_t>::max());
+  // Check for overflow into workerId_ section.
+  TORCH_INTERNAL_ASSERT(next_autograd_message_id_ < max_id_);
+  return next_autograd_message_id_++;
+}
+
+DistAutogradContext& DistAutogradContainer::createContext(int64_t context_id) {
+  checkInitialized();
+
+  std::lock_guard<std::mutex> guard(autograd_context_lock_);
+  auto it = autograd_context_.find(context_id);
+  if (it != autograd_context_.end()) {
+    return it->second;
   }
 
+  auto& context = autograd_context_
+                      .emplace(
+                          std::piecewise_construct,
+                          std::forward_as_tuple(context_id),
+                          std::forward_as_tuple(context_id))
+                      .first->second;
+  current_context_id_ = context_id;
+  return context;
+}
+
+void DistAutogradContainer::checkInitialized() {
+  TORCH_CHECK(
+      initialized_,
+      "Need to initialize distributed autograd using "
+      "torch.distributed.autograd.init()");
+}
+
+const DistAutogradContext& DistAutogradContainer::newContext() {
+  checkInitialized();
+
   std::lock_guard<std::mutex> guard(autograd_context_lock_);
-  TORCH_INTERNAL_ASSERT(
-      next_context_id_ < std::numeric_limits<int64_t>::max() &&
-          next_context_id_ <
-              (kMaxContextId |
-               (static_cast<int64_t>(worker_id_) << kContextIdBits)),
-      "We have run out of autograd context ids!!!");
-
-  autograd_context_.emplace(
-      std::piecewise_construct,
-      std::forward_as_tuple(next_context_id_),
-      std::forward_as_tuple(next_context_id_));
-
-  current_context_id_ = next_context_id_;
-  return autograd_context_.at(next_context_id_++);
+  // Check for int64_t max overflow.
+  TORCH_INTERNAL_ASSERT(next_context_id_ < std::numeric_limits<int64_t>::max());
+  // Check for overflow into workerId_ section.
+  TORCH_INTERNAL_ASSERT(next_context_id_ < max_id_);
+
+  auto& context = autograd_context_
+                      .emplace(
+                          std::piecewise_construct,
+                          std::forward_as_tuple(next_context_id_),
+                          std::forward_as_tuple(next_context_id_))
+                      .first->second;
+
+  current_context_id_ = next_context_id_++;
```

Comment:
when exiting a distributed autograd context, are we setting `current_context_id_` back to -1?